util/piggy_prepare.py

```diff
@@ -0,0 +1,339 @@
+import hashlib
+import inspect
+import json
+import os
+import sys
+import time
+
+
+def piggy_prepare_run(dir, script, source, target, is_pgxl):
+
+    date = time.strftime('%Y%m%d')
+    fp = date + '_' + hashlib.sha1('\t'.join([dir, script, source, target])).hexdigest()[:8]
+    fname = 'piggy_func_' + fp
+    sview = 'piggy_source_' + fp
+    tview = 'piggy_target_' + fp
+
+    if ' ' not in source:
+        source_query = 'SELECT * FROM ' + source
+    else:
+        source_query = source
+
+    if '(' not in target:
+        target_query = 'SELECT * FROM ' + target
+    else:
+        tname, tcols = target.strip(' )').split('(')
+        target_query = 'SELECT %s FROM %s' % (tcols, tname)
+
+    logfile_path = os.path.join(dir, fname + '.log')
+    if os.path.isfile(logfile_path):
+        os.remove(logfile_path)
+
+    ts_format = '%m-%d %H:%M:%S'
+
+    # Why TABLE not VIEW?
+    # Because PGXL does not create views on data nodes.
+
+    sql_create_tables = '''
+    DROP TABLE IF EXISTS %(sview)s CASCADE;
+    DROP TABLE IF EXISTS %(tview)s CASCADE;
+    CREATE TABLE %(sview)s AS SELECT * FROM (%(source_query)s) _sview_source LIMIT 0;
+    CREATE TABLE %(tview)s AS %(target_query)s LIMIT 0;
+    ''' % {
+        'sview': sview,
+        'tview': tview,
+        'source_query': source_query,
+        'target_query': target_query,
+    }
+
+
+    # Why log to file instead of a DB table?
+    # PGXL would complain something like 'no XID'.
+    # This may have fixed it though: https://github.com/snaga/postgres-xl/commit/d4136935a1d741c61ada13ef8c5ae44f68162cc9
+
+    sql_create_functions = '''
+    CREATE OR REPLACE FUNCTION %(fname)s (record %(sview)s)
+        RETURNS SETOF %(tview)s
+    AS $func$
+
+        # NULL input tuple indicates end of stream.
+        # It's time to flush output and clean up.
+        if record is None:
+            import os
+            if 'piggy' not in SD:
+                raise StopIteration
+            piggy = SD['piggy']
+            try:
+                piggy.log('Grabbing final tuples from the buffer.')
+                for x in piggy.pull_final():
+                    yield x
+            finally:
+                del SD['piggy']
+            raise StopIteration
+
+        # First input tuple; we setup the worker process.
+        if 'piggy' not in SD:
+            import errno
+            import fcntl
+            import json
+            import os
+            import socket
+            import subprocess
+            import time
+            from datetime import datetime
+            from Queue import Queue, Empty, Full
+            from threading import Thread
+
+            # Create a pipe for logs.
+            # Only one process would succeed on the same host.
+            log_pipe_name = os.path.join('%(dir)s', '%(fname)s.pipe')
+            try:
+                os.mkfifo(log_pipe_name)
```

Comment:
@raphaelhoffmann and I had a chat about the SSH requirement and we concluded it doesn't make sense to support the case where the database nodes disallow SSH access, because the DD user will most likely be the one who sets up the database as well. (We're making it easier so this is plausible.)  Therefore, the user has to have SSH access to even start the datanode on such host. (Sharing a parallel database between users never worked in practice anyway, let alone the case where a single user tries to run multiple queries..)

SSH authentication issues can always be handled transparently with ssh-agent or Kerberos.  Adding additional config params to DD will only add unnecessary complexity and I know such approach always break at some point.

Anyway, so is it possible to fix the issue with minimal change?